internal/graph/cursors.go

```diff
@@ -310,3 +324,197 @@ func combineCursors(cursor *v1.Cursor, toAdd *v1.Cursor) (*v1.Cursor, error) {
+// emitted to the parent stream, while subsequent tasks' results are emitted in the defined order of the tasks
+// to ensure cursors and limits work as expected.
+type parallelLimitedIndexedStream[Q any] struct {
+	lock sync.Mutex
+
+	ctx          context.Context
+	ci           cursorInformation
+	parentStream dispatch.Stream[Q]
+
+	streamCount      int
+	toEmitIndex      int
+	countingStream   *dispatch.CountingDispatchStream[Q]
+	childStreams     map[int]*dispatch.CollectingDispatchStream[Q]
+	completedIndexes map[int]bool
+}
+
+func newParallelLimitedIndexedStream[Q any](
+	ctx context.Context,
+	ci cursorInformation,
+	parentStream dispatch.Stream[Q],
+	streamCount int,
+) (*parallelLimitedIndexedStream[Q], error) {
+	if streamCount <= 0 {
+		return nil, spiceerrors.MustBugf("got invalid stream count")
+	}
+
+	return &parallelLimitedIndexedStream[Q]{
+		ctx:              ctx,
+		ci:               ci,
+		parentStream:     parentStream,
+		countingStream:   nil,
+		childStreams:     map[int]*dispatch.CollectingDispatchStream[Q]{},
+		completedIndexes: map[int]bool{},
+		toEmitIndex:      0,
+		streamCount:      streamCount,
+	}, nil
+}
+
+// forTaskIndex returns a new context, stream and cursor for invoking the task at the specific index and publishing its results.
+func (ls *parallelLimitedIndexedStream[Q]) forTaskIndex(ctx context.Context, index int, currentCursor cursorInformation) (context.Context, dispatch.Stream[Q], cursorInformation) {
+	// Create a new cursor with cloned limits, because each child task which executes (in parallel) will need its own
+	// limit tracking. The overall limit on the original cursor is managed in completedTaskIndex.
+	childCI, cctx := currentCursor.withClonedLimits(ctx)
+
+	// If executing for the first index, it can stream directly to the parent stream, but we need to count the number
+	// of items streamed to adjust the overall limits.
+	if index == 0 {
+		countingStream := dispatch.NewCountingDispatchStream[Q](ls.parentStream)
+		ls.countingStream = countingStream
+		return cctx, countingStream, childCI
+	}
+
+	// Otherwise, create a child stream with an adjusted limits on the cursor. We have to clone the cursor's
+	// limits here to ensure that the child's publishing doesn't affect the first branch.
+	ls.lock.Lock()
+	defer ls.lock.Unlock()
+
+	childStream := dispatch.NewCollectingDispatchStream[Q](ctx)
+	ls.childStreams[index] = childStream
+	return cctx, childStream, childCI
+}
+
+// completedTaskIndex indicates the the task at the specific index has completed successfully and that its collected
+// results should be published to the parent stream, so long as all previous tasks have been completed and published as well.
+func (ls *parallelLimitedIndexedStream[Q]) completedTaskIndex(index int) error {
+	ls.lock.Lock()
+	defer ls.lock.Unlock()
+
+	// Mark the task as completed.
+	ls.completedIndexes[index] = true
```

Comment:
Let's say stream 3 completes before stream 2, how's stream 3 going to emit results if it's waiting for stream 2, but stream 2 cannot be marked as completed because of the mutex being acquired by 3?

Reply:
*stream* 3 is marked as completed; but it is not marked as published; `completed` is for the task, but publishing happens after it